### PR TITLE
Fix average_neighbor_degree calculations for directed graph

### PR DESCRIPTION
**Pull request by yusuf-csdev**
Fixes: #5383

There was a problem with average neighbor degree calculations in a directed graph. The method was working as expected when the source argument was "out". But when the source argument was "in", it gave different results then expected. 

I have made it so that we check the source argument before deciding on which nodes are neighbors. 
Let's say we have the following directed graph: 1->2->3->4
and we are asked to find the average neighbor degree for node 2. Then if the source argument is "in", we should look for the nodes which have an edge "to" node 2 and these will be neighbors of node 2. However, if the source parameter is "out", we should look for the nodes which have an edge "from" node 2 and these will be neighbors of node 2.

I have also updated the documentation of method. I ran tests and some of the results were different from the original test results as expected. I used pen and paper, and checked the work I have done. After getting expected results, I updated test_neighbor_degree file and changed assert statements.

### CODE REVIEW

File: networkx/algorithms/assortativity/neighbor_degree.py
Comment (dschult) [inline networkx/algorithms/assortativity/neighbor_degree.py:162]: Can this use `G.pred[n]` or `G.succ[n]` instead of `G[n]`?
It feels like we shouldn't have to rebuild the adjacency dict.

Also, how do we handle the "in+out" case? `G[n]` only yields outgoing neighbors. 
So we need a way to chain the incoming and outgoing neighbors, while making sure
to handle a neighbor that is both incoming and outgoing appropriately.

Comment (dschult) [reply networkx/algorithms/assortativity/neighbor_degree.py:162]: If you get an error about not finding `G.succ` for class `Graph`, then you are treating an undirected graph as a directed graph somewhere. Please track that down. Only DiGraph and MultiDiGRaph have the attributes `G.succ` and `G.pred`.

I like your approach of `P_n` and `S_n` along with making them empty dicts when they aren't relevant. 

Two quick comments... Can you use # for comments instead of """?  And is there a way to pull the if statements out of the loop so that logic can be done once before the loop instead. Something like:
```python
    # default to empty neighbors dicts
    G_P = S_P = {n: {} for n in G}
    if G.is_directed():
        if "in" in source:
            G_P = G.pred
        if "out" in source:
            G_S = G.succ
    else:
        G_S = G.adj
```
Then inside the loop you can use `P_n = G_P[n]` and `S_n = G_S[n]`

Comment (yusuf-csdev) [reply networkx/algorithms/assortativity/neighbor_degree.py]: Thank you for your comments. I have updated the code accordingly and committed the new version. If there are any other issues, it would be great to learn and solve them.

**Comment by dschult** [comment]
I think this looks good. The docstring lines and comment lines are sometimes longer than the 88 char enforced by black and so could be shortened. I think some of the comment lines aren't needed because the code says what the comment says in a more compact way.  But these are style issues.  

What I'm waiting for before I approve this is a concensus on whether we can just change this and put a note in the release notes that the values of `average_neighbor_degree` have changed for DiGraphs, or whether we need to deprecate the old version in some way.  I would argue that the old values for DiGraph didn't make sense, so we should change it (correct it?) now... not waiting for a deprecation cycle.  @rossbar do you have some insight on deprecation protocol here?

**Comment by rossbar** [comment]
>  I would argue that the old values for DiGraph didn't make sense, so we should change it (correct it?) now... not waiting for a deprecation cycle

Agreed - this seems like it's fixing "defective" behavior, so I'd say no warnings necessary. A mention in the release note (potentially with an example) would be good :+1:

**Comment by dschult** [comment]
Great!  Thanks @rossbar!
@yusuf-csdev can you come up with a simple example (2-5 lines?) that shows how this changes the computed values? We should put that into `doc/release/release_dev.rst`.  I can do this if you prefer <especially since I will likely edit it anyway>... :}
Thoughts?

**Comment by dschult** [comment]
I went ahead and added the release notes.  They are probably kind of long, but can be edited down when the release is made if desired.

I also tried to shrink the docstring line-length and comments. See what you think @yusuf-csdev 
Because I pushed to your fork (the PR), you'll have to pull from your fork before making further changes to this branch.

I did change the code in minor ways -- I moved `arg = {}` to just before the loop where it is computed. And I realized that
S_n and P_n were only used once, so I just replaced them with `G_S[n]` and `G_P[n]`. Then for spacing reasons I renamed `tgt_deg` to `t_deg`.

Feel free to change anything you like back to the way it was or to something even better.

**Comment by yusuf-csdev** [comment]
@dschult Thank you for the changes you have made. They all look good. If I am not mistaken, you have added an example to the release notes. If you want me to an extra one besides that, please let me know.

**Review by rossbar — APPROVED** [review]
Thanks @yusuf-csdev and @dschult , this LGTM especially with the updated docs and detailed release note!

**Review by dschult — APPROVED** [review]

**Pull request merged**
